mpy-cross/mpy_cross/compiler.py

```diff
@@ -0,0 +1,204 @@
+# This file is part of the MicroPython project, http://micropython.org/
+#
+# The MIT License (MIT)
+#
+# Copyright (c) 2022 Andrew Leech
+# Copyright (c) 2023 Jim Mussared
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import os
+
+
+class CrossCompileError(Exception):
+    pass
+
+
+class CompilerNotFoundError(Exception):
+    pass
+
+
+EMIT_BYTECODE = "bytecode"
+EMIT_NATIVE = "native"
+EMIT_VIPER = "viper"
+
+
+def emit_id(name):
+    return {
+        EMIT_BYTECODE: 1,
+        EMIT_NATIVE: 2,
+        EMIT_VIPER: 3,
+    }[name]
+
+
+NATIVE_ARCHS = {
+    "NATIVE_ARCH_NONE": "",
+    "NATIVE_ARCH_X86": "x86",
+    "NATIVE_ARCH_X64": "x64",
+    "NATIVE_ARCH_ARMV6": "armv6",
+    "NATIVE_ARCH_ARMV6M": "armv6m",
+    "NATIVE_ARCH_ARMV7M": "armv7m",
+    "NATIVE_ARCH_ARMV7EM": "armv7em",
+    "NATIVE_ARCH_ARMV7EMSP": "armv7emsp",
+    "NATIVE_ARCH_ARMV7EMDP": "armv7emdp",
+    "NATIVE_ARCH_XTENSA": "xtensa",
+    "NATIVE_ARCH_XTENSAWIN": "xtensawin",
+}
+
+globals().update(NATIVE_ARCHS)
+
+
+def native_arch_id(name):
+    if name is None:
+        return 0
+
+    return {
+        "": 0,
+        "x86": 1,
+        "x64": 2,
+        "armv6": 3,
+        "armv6m": 4,
+        "armv7m": 5,
+        "armv7em": 6,
+        "armv7emsp": 7,
+        "armv7emdp": 8,
+        "xtensa": 9,
+        "xtensawin": 10,
+    }[name]
+
+
+class Compiler:
+    def mpy_version(self):
+        """
+        Get the version and sub-version of the .mpy file format generated by this compiler.
+
+        Returns: A tuple of `(mpy_version, mpy_sub_version)`
+        """
+        raise NotImplementedError()
+
+    def version(self):
+        """
+        Get the version string from this compiler.
+
+        Returns: A string, typically "MicroPython <git tag> on <build date>".
+        """
+        raise NotImplementedError()
+
+    def compile(self, src, dest=None, src_path=None, opt=0, march=None, emit=EMIT_BYTECODE):
```

Comment:
The `compile` method already exists (it's used in makemanifest.py and is supported in the existing mpy_cross package on PyPI).

However, what if we allowed `src` and `dest` to be file-like objects? (In particular, I'm thinking `io.BytesIO`). Would that suit your use case?

Reply:
Yes, I think so. I'm assuming it could work like this...

If `src` is a string, pass it as a command line arg or if it is a file-like object, pipe it to stdin of the `mpy-cross` process. Likewise, if `dest` is a string, pass it as the `--output` argument or if it is a file-like object, pip it to stdout of the `mpy-cross` process.